[Cloudflare] Support limited-scope API tokens

luckyrat opened the pull request:
This improves the security of Cloudflare accounts by allowing the use of an API token that is limited in scope to only the permissions required to modify a particular DNS zone.

It may or may not be obvious that this is the first time I've contributed any Rust code and I've yet to have time to learn the idioms and best-practices of the language. Hence there may be a number of aspects of this which you're unhappy with - I'm willing to make some changes if you'd like and equally OK if you just take this as an inspiration for your own entirely different implementation.

I'm particularly unsure how best to handle the config file changes. At the moment there's not really any validation that the config is in a meaningful state (i.e. that the correct combination of token, key and email are supplied) so all 3 have to be defined, even if the unused ones are empty strings.

This version 4.0.0-pre is running on one of my servers and performing well so far. Thanks for developing this useful application!

nickbabcock commented on src/cloudflare.rs:
```diff
@@ -117,18 +118,34 @@ impl fmt::Display for ClError {
+    email: String,
+    key: String,
+    token: String
+) -> reqwest::RequestBuilder {
```
This can be written a bit better

```rust
fn apply_auth_to_request(
    request_builder: reqwest::RequestBuilder,
    email: String,
    key: String,
    token: String,
) -> reqwest::RequestBuilder {
    if !token.is_empty() {
        request_builder.bearer_auth(token)
    } else {
        request_builder
            .header("X-Auth-Email", email)
            .header("X-Auth-Key", key)
    }
}
```

Hmm wish I had a bit better design so that it wouldn't be necessary to pass in so many parameters (but you're good 👍 )

luckyrat replied:
Yeah my first attempt was passing in the config object (if that's the right terminology for Rust!) but the need to support the use of the CloudflareClient object too made me switch to multiple parameters. I'm not quite ready for writing a Rust function with a parameter that can be one of multiple types - no idea if the idea even makes any sense in the language so I figured this would be good enough for the time being :-)

Thanks for the Rust tips. I've made the change as you suggested.

nickbabcock commented:
This is a must have PR and using tokens should definitely be the recommended way forward. I think this PR is excellently written with just a few comments:

No need to update the version found in Cargo.toml

> I'm particularly unsure how best to handle the config file changes. At the moment there's not really any validation that the config is in a meaningful state (i.e. that the correct combination of token, key and email are supplied) so all 3 have to be defined, even if the unused ones are empty strings.

I'm not too concerned about addressing config validations in this PR. It can be addressed in a subsequent PR before a release. We may not need to get too fancy at this stage -- a warning that the supplied email and key will go unused when token is provided as well may be sufficient. An error for when an email is supplied but not a key, etc.

I noticed that you added yourself to the authors field. Did you want to take on an active role in dness (ie: add features / fix bugs / issue triaging)? If so, great! I'll add you as a collaborator to the repo. Else it can be omitted from the PR to avoid sending you unnecessary pings.

luckyrat commented:
> No need to update the version found in Cargo.toml

I think that's the only way to get a different version number in the deb package output? In any case, that was only needed during my development and testing so if you or the CI will handle the version change in future, that's fine with me. I've reverted the change now.

> I noticed that you added yourself to the authors field. Did you want to take on an active role in dness (ie: add features / fix bugs / issue triaging)? If so, great! I'll add you as a collaborator to the repo. Else it can be omitted from the PR to avoid sending you unnecessary pings.

I only did that for the purpose of transparent audit tracking of who the contributors to the package have been but I wasn't sure it was the right way to use Cargo metadata. It sounds like it's not for the purpose I imagined so I've removed myself from that list.

Thanks for the offer of collaborator status. Given my lack of experience with Rust and a lack of time to learn it thoroughly right now, I'll take a non-committal approach to collaboration for the time being. I'll definitely keep an eye on the project though and might chip in with some more PRs and issue discussion occasionally.

I'll take a look into that merge conflict now...

luckyrat commented:
Looks like the merge conflict is with the PATCH patch you just merged into master. Would you like me to rebase this branch onto master, push a merge commit to this branch or just leave you to resolve it?

nickbabcock commented:
Rebase please 😄 

> I think that's the only way to get a different version number in the deb package output? In any case, that was only needed during my development and testing so if you or the CI will handle the version change in future, that's fine with me. I've reverted the change now.

Yeah no worries, that's how I test as well, just no need to include in the commit.

> Thanks for the offer of collaborator status. Given my lack of experience with Rust and a lack of time to learn it thoroughly right now, I'll take a non-committal approach to collaboration for the time being. I'll definitely keep an eye on the project though and might chip in with some more PRs and issue discussion occasionally.

Yeah np. It probably comes down to semantics, but author (at least to me) comes with a heavier weight than contributor

luckyrat commented:
I've just pushed the rebase so we should be good to go once CI passes. Ping me if you have any other queries before merging :-)

nickbabcock commented on src/cloudflare.rs:
```diff
@@ -297,11 +317,13 @@ impl<'a> CloudflareClient<'a> {
+
+            request_builder = apply_auth_to_request(request_builder, self.email.clone(), self.key.clone(), self.token.clone());
+                
+            let response: CloudflareResponse<Vec<CloudflareDnsRecord>> = request_builder
```
I believe:

```
CloudflareResponse<Vec<CloudflareDnsRecord>>
```

should be

```
CloudflareResponse<CloudflareDnsRecord>
```

I receive an error trying to interpret cloudflare's response (the update is still effective, it's just that an error is printed on updates).

luckyrat replied:
Oops. Yeah you're right. I've pushed a fixed version.

nickbabcock replied:
Excellent, run `cargo fmt` in the repo to format the code and this PR will look to be all set to be merged. I tested it out and it worked well 🎉

luckyrat replied:
Great. I've done that and pushed a new commit since the format changes probably spanned across multiple commits and I'm not sure if you just intend to squash them all upon merge anyway. Let me know if you'd like me to take an alternative approach.

nickbabcock replied:
You're good. My plan is to cut a patch release with the current master, merge this PR, touch things up a bit, and then cut another release in a week or so. Thanks again
